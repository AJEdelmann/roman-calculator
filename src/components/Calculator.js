import React, { useState, useEffect } from "react";
import { validateNumber, toRoman, toNumber } from "../utils/convert";
import Button from "./Button";
import Display from "./Display";

const Calculator = () => {
  const [input, setInput] = useState("");
  const [error, setError] = useState("");
  const [curNum, setCurNum] = useState("");
  const [prevNum, setPrevNum] = useState("");
  const [result, setResult] = useState("");
  const [operator, setOperator] = useState("");
  const [history, setHistory] = useState("");

  // Data for rendering the buttons
  const numbersKeys = ["C", "D", "M", "V", "X", "L", "CE", "I", "="];
  const operatorsKeys = ["+", "-", "Ã—"];

  const handleInput = val => {
    // To clean the display automatically if you start typing a number after an operation
    if (result && result === input) {
      setInput("");
    }
    setInput(prevInput => prevInput + val);
  };

  const clearInput = () => {
    setInput("");
    setHistory("");
  };

  // Input manipulation triggered after user click an operator
  const calculation = operator => {
    setPrevNum(input);
    setInput("");
    setOperator(operator);
    setHistory(`${history} ${input} ${operator} `);
  };

  const evaluate = () => {
    const prevNumber = toNumber(prevNum);
    const curNumber = toNumber(curNum);
    let operation;

    if (operator === "+") {
      operation = toRoman(prevNumber + curNumber);
    } else if (operator === "-") {
      operation = toRoman(prevNumber - curNumber);
    } else if (operator === "Ã—") {
      operation = toRoman(prevNumber * curNumber);
    } else {
      // In case the user click the equal button without previously selecting an operator
      operation = toRoman(curNumber);
    }
    setResult(operation);
    setInput(operation);
    setOperator("");
    setHistory(`${history} ${input}`);
  };

  useEffect(() => {
    setError(validateNumber(input));
    setCurNum(input);
  }, [input]);

  return (
    <div className="calculator instructions-bounce ">
      <Display error={error} history={history}>
        {input}
      </Display>

      <div className="numbers-wrapper">
        {numbersKeys.map(k => {
          return (
            <Button
              data={k}
              key={k}
              error={k !== "CE" ? error : null}
              handleClick={
                k === "CE" ? clearInput : k === "=" ? evaluate : handleInput
              }
              className={`button ${
                k === "=" || k === "CE" ? "utility" : "number"
              } ${k === "CE" ? "ce" : ""}`}
            >
              {k}
            </Button>
          );
        })}
      </div>
      <div className="operators-wrapper">
        {operatorsKeys.map(k => {
          return (
            <Button
              key={k}
              error={error}
              className="button operator"
              handleClick={calculation}
              data={k}
            ></Button>
          );
        })}
      </div>
    </div>
  );
};

export default Calculator;
