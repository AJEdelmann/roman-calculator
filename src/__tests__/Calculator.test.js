import React from "react";
import ReactDOM from "react-dom";
import Calculator from "../components/Calculator";
import { cleanup } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import renderer from "react-test-renderer";

afterEach(cleanup);

describe("<Calculator/>", () => {
  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<Calculator />, div);
  });

  it("matches the snapshot", () => {
    const tree = renderer.create(<Calculator />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
