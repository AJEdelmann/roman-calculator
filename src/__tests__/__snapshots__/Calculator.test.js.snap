// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Calculator/> matches the snapshot 1`] = `
<div
  className="calc-wrapper"
>
  <div
    className="display-wrapper  "
  >
    <div
      className="display-error-message"
    >
      
    </div>
    <div
      className="display-number"
    >
      
    </div>
  </div>
  <div
    className="numbers-wrapper"
  >
    <div
      className="button number "
      onClick={[Function]}
      style={null}
    >
      <p
        className="p-operator"
      >
        C
      </p>
    </div>
    <div
      className="button number "
      onClick={[Function]}
      style={null}
    >
      <p
        className="p-operator"
      >
        D
      </p>
    </div>
    <div
      className="button number "
      onClick={[Function]}
      style={null}
    >
      <p
        className="p-operator"
      >
        M
      </p>
    </div>
    <div
      className="button number "
      onClick={[Function]}
      style={null}
    >
      <p
        className="p-operator"
      >
        V
      </p>
    </div>
    <div
      className="button number "
      onClick={[Function]}
      style={null}
    >
      <p
        className="p-operator"
      >
        X
      </p>
    </div>
    <div
      className="button number "
      onClick={[Function]}
      style={null}
    >
      <p
        className="p-operator"
      >
        L
      </p>
    </div>
    <div
      className="button utility ce"
      onClick={[Function]}
      style={null}
    >
      <p
        className="p-operator"
      >
        CE
      </p>
    </div>
    <div
      className="button number "
      onClick={[Function]}
      style={null}
    >
      <p
        className="p-operator"
      >
        I
      </p>
    </div>
    <div
      className="button utility "
      onClick={[Function]}
      style={null}
    >
      <p
        className="p-operator"
      >
        =
      </p>
    </div>
  </div>
  <div
    className="operators-wrapper"
  >
    <div
      className="button operator"
      onClick={[Function]}
      style={null}
    >
      <p
        className="p-operator"
      >
        +
      </p>
    </div>
    <div
      className="button operator"
      onClick={[Function]}
      style={null}
    >
      <p
        className="p-operator"
      >
        -
      </p>
    </div>
    <div
      className="button operator"
      onClick={[Function]}
      style={null}
    >
      <p
        className="p-operator"
      >
        Ã—
      </p>
    </div>
  </div>
</div>
`;
